src/uiHelpers_nanos.c

```diff
@@ -0,0 +1,212 @@
+        unsigned int button_mask_counter MARK_UNUSED
+)
+{
+	return 0;
+}
+
+static void scroll_update_display_content()
+{
+	scrollingState_t* ctx = scrollingState;
+	assert_uiScrolling_magic();
+	ASSERT(ctx->currentText[SIZEOF(ctx->currentText) - 1] == '\0');
+	ASSERT(ctx->scrollIndex + SIZEOF(ctx->currentText) <= SIZEOF(ctx->fullText));
+	os_memmove(
+	        ctx->currentText,
+	        ctx->fullText + ctx->scrollIndex,
+	        SIZEOF(ctx->currentText) - 1
+	);
+	UX_REDISPLAY();
+}
+
+static void scroll_left()
+{
+	scrollingState_t* ctx = scrollingState;
+	assert_uiScrolling_magic();
+	if (ctx->scrollIndex > 0) {
+		ctx->scrollIndex--;
+		scroll_update_display_content();
+	}
+}
+
+static void scroll_right()
+{
+	scrollingState_t* ctx = scrollingState;
+	assert_uiScrolling_magic();
+	if (ctx->scrollIndex + SIZEOF(ctx->currentText) < 1 + strlen(ctx->fullText)) {
+		scrollingState->scrollIndex++;
+		scroll_update_display_content();
+	}
+}
+
+unsigned int ui_scrollingText_button(
+        unsigned int button_mask,
+        unsigned int button_mask_counter MARK_UNUSED
+)
+{
+	scrollingState_t* ctx = scrollingState;
+	TRY_CATCH_UI({
+		assert_uiScrolling_magic();
+		ASSERT(io_state == IO_EXPECT_UI);
+		ASSERT(device_is_unlocked() == true);
+		switch (button_mask)
+		{
+		case BUTTON_LEFT:
+		case BUTTON_EVT_FAST | BUTTON_LEFT: // SEEK LEFT
+			scroll_left();
+			break;
+
+		case BUTTON_RIGHT:
+		case BUTTON_EVT_FAST | BUTTON_RIGHT: // SEEK RIGHT
+			scroll_right();
+			break;
+
+		case BUTTON_EVT_RELEASED | BUTTON_LEFT | BUTTON_RIGHT: // PROCEED
+			uiCallback_confirm(&ctx->callback);
+			break;
+		}
+	});
+	return 0;
+}
+
+unsigned int ui_confirm_button(
+        unsigned int button_mask,
+        unsigned int button_mask_counter MARK_UNUSED
+)
+{
+	TRY_CATCH_UI({
+		assert_uiConfirm_magic();
+		ASSERT(io_state == IO_EXPECT_UI);
+		ASSERT(device_is_unlocked() == true);
+		switch (button_mask)
+		{
+		case BUTTON_EVT_RELEASED | BUTTON_LEFT: // REJECT
+			uiCallback_reject(&confirmState->callback);
+			break;
+
+		case BUTTON_EVT_RELEASED | BUTTON_RIGHT: // APPROVE
+			uiCallback_confirm(&confirmState->callback);
+			break;
+		}
+	})
+	return 0;
+}
+
+static const bagl_element_t ui_scrollingText[] = {
+	UI_BACKGROUND(),
+	UI_ICON_LEFT(ID_ICON_GO_LEFT, BAGL_GLYPH_ICON_LEFT),
+	UI_ICON_RIGHT(ID_ICON_GO_RIGHT, BAGL_GLYPH_ICON_RIGHT),
+
+	// TODO(ppershing): what are the following magical numbers?
+
+	UI_TEXT(ID_UNSPECIFIED, 0, 12, 128, &displayState.scrolling.header),
+	UI_TEXT(ID_UNSPECIFIED, 0, 26, 128, &displayState.scrolling.currentText),
+	#ifdef HEADLESS
+	HEADLESS_UI_ELEMENT(),
+	#endif
+};
+
+static const bagl_element_t* ui_prepro_scrollingText(const bagl_element_t *element)
+{
+	scrollingState_t* ctx = scrollingState;
+	assert_uiScrolling_magic();
+
+	switch (element->component.userid) {
+	case ID_ICON_GO_LEFT:
+		return (ctx->scrollIndex == 0 && SIZEOF(ctx->currentText) < strlen(ctx->fullText))
```

Comment:
sizeof -> strlen. (there is an invariant that strlen(x) + 1 <= sizeof(x)) so this condition is off by one